src/common/substrait/src/df_expr.rs

```diff
@@ -0,0 +1,731 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::VecDeque;
+use std::str::FromStr;
+
+use datafusion::logical_plan::{Column, Expr};
+use datafusion_expr::{expr_fn, BuiltinScalarFunction, Operator};
+use datatypes::schema::Schema;
+use snafu::{ensure, OptionExt};
+use substrait_proto::protobuf::expression::field_reference::ReferenceType as FieldReferenceType;
+use substrait_proto::protobuf::expression::reference_segment::{
+    ReferenceType as SegReferenceType, StructField,
+};
+use substrait_proto::protobuf::expression::{
+    FieldReference, ReferenceSegment, RexType, ScalarFunction,
+};
+use substrait_proto::protobuf::function_argument::ArgType;
+use substrait_proto::protobuf::Expression;
+
+use crate::context::ConvertorContext;
+use crate::error::{
+    EmptyExprSnafu, InvalidParametersSnafu, MissingFieldSnafu, Result, UnsupportedExprSnafu,
+};
+
+/// Convert substrait's `Expression` to DataFusion's `Expr`.
+pub fn to_df_expr(ctx: &ConvertorContext, expression: Expression, schema: &Schema) -> Result<Expr> {
+    let expr_rex_type = expression.rex_type.context(EmptyExprSnafu)?;
+    match expr_rex_type {
+        RexType::Literal(_) => UnsupportedExprSnafu {
+            name: "substrait Literal expression",
+        }
+        .fail()?,
+        RexType::Selection(selection) => convert_selection_rex(*selection, schema),
+        RexType::ScalarFunction(scalar_fn) => convert_scalar_function(ctx, scalar_fn, schema),
+        RexType::WindowFunction(_)
+        | RexType::IfThen(_)
+        | RexType::SwitchExpression(_)
+        | RexType::SingularOrList(_)
+        | RexType::MultiOrList(_)
+        | RexType::Cast(_)
+        | RexType::Subquery(_)
+        | RexType::Enum(_) => UnsupportedExprSnafu {
+            name: format!("substrait expression {:?}", expr_rex_type),
+        }
+        .fail()?,
+    }
+}
+
+/// Convert Substrait's `FieldReference` - `DirectReference` - `StructField` to Datafusion's
+/// `Column` expr.
+pub fn convert_selection_rex(selection: FieldReference, schema: &Schema) -> Result<Expr> {
+    if let Some(FieldReferenceType::DirectReference(direct_ref)) = selection.reference_type
+    && let Some(SegReferenceType::StructField(field))  = direct_ref.reference_type{
+        let column_name = schema.column_name_by_index(field.field as _).to_string();
+         Ok(Expr::Column(Column{ relation: None, name: column_name }))
+    } else {
+        InvalidParametersSnafu{reason:"Only support direct struct reference in Selection Rex"}.fail()
+    }
```

Comment:
Consider using the `if-let` statement which rustfmt supports to format. The following codes are formatted:
```suggestion
    if let Some(FieldReferenceType::DirectReference(direct_ref)) = selection.reference_type {
        if let Some(SegReferenceType::StructField(field)) = direct_ref.reference_type {
            let column_name = schema.column_name_by_index(field.field as _).to_string();
            return Ok(Expr::Column(Column {
                relation: None,
                name: column_name,
            }));
        }
    }

    InvalidParametersSnafu {
        reason: "Only support direct struct reference in Selection Rex",
    }
    .fail()
```